corehq/apps/change_feed/pillow.py

```diff
@@ -26,19 +27,20 @@ def __init__(self, kafka, data_source_type, data_source_name):
         self._data_source_name = data_source_name
 
     def process_change(self, pillow_instance, change, do_set_checkpoint=False):
-        document_type = _get_document_type(change.document)
-        if document_type:
-            assert change.document is not None
-            change_meta = ChangeMeta(
-                document_id=change.id,
+        try:
+            doc_meta = get_doc_meta_object_from_document(change.document)
+            change_meta = change_meta_from_doc_meta_and_document(
+                doc_meta=doc_meta,
+                document=change.document,
                 data_source_type=self._data_source_type,
                 data_source_name=self._data_source_name,
-                document_type=document_type,
-                document_subtype=_get_document_subtype(change.document),
-                domain=change.document.get('domain', None),
-                is_deletion=change.deleted,
+                doc_id=change.id,
             )
-            self._producer.send_change(get_topic(document_type), change_meta)
+        except MissingMetaInformationError:
+            pass
+        else:
+            assert not change.deleted
```

Comment:
yeah this is super confusing. on a couch feed `change.deleted` will be true if it's a hard deletion in couch. however nothing in the framework currently supports hard deletions because we can't infer anything about the document since we don't have the type/domain/etc. (unless we go digging through deleted revisions which we could consider doing in the future).

on soft deletions `change_meta.is_deletion` will be set properly (e.g. [here](https://github.com/dimagi/commcare-hq/pull/11210/files#diff-10236d523751301be50e82cf32d9e118R103) and then the kafka change feed will pass that up to `change.deleted` [here](https://github.com/dimagi/commcare-hq/blob/master/corehq/apps/change_feed/consumer/feed.py#L167).

This can definitely be cleaned up but I think that might be easier once ~everything is running off the kafka feeds